### DISCUSSION TITLE
How to "environmentalize" a workflow?

### DISCUSSION THREAD

**Discussion by mellis481** [community/community#25776] [Actions]
<p>I’m trying to understand how you can “environmentalize” a Github actions workflow, that is, have a workflow use different values depending on the environment.  Here is my use-case:</p>
<ul>
<li>I have a manual workflow to deploy my application to an environment (development, qa, staging, production) which is specified via a workflow input.</li>
<li>The environments are hosted in AWS and AWS credentials a different for each environment.</li>
</ul>
<p>Configuring <em>an</em> environment is as simple as this:</p>
<pre><code class="lang-auto">- name: Configure AWS credentials
  uses: aws-actions/configure-aws-credentials@v1
  with:
    aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID_DEVELOPMENT }}
    aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY_DEVELOPMENT }}
    aws-region: 'us-east-1'
</code></pre>
<p><strong>My question is what is the best way to have my workflow ensure all environments will be configured.</strong></p>
<p>I’ve done a bit of searching for this answer, but haven’t found anything.  The only option I’ve seen so far is to use Github actions contexts which would result in my workflow having a step for each environment where i have an <code>if</code> expression to determine which step to actually run.  Eg.</p>
<pre><code class="lang-auto">- name: Configure Dev AWS credentials
  if: ${{ github.event.inputs.env_to_promote_to === 'development' }}
  uses: aws-actions/configure-aws-credentials@v1
  with:
    aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID_DEVELOPMENT }}
    aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY_DEVELOPMENT }}
    aws-region: 'us-east-1'
</code></pre>
<p>That seems like a really suboptimal solution since most of my *.yml file will be conditional steps to configure AWS.</p>
<p><a href="https://github.com/actions/starter-workflows/issues/68#issuecomment-552074596" rel="noopener nofollow ugc">This</a> discussion seemed to offer a potential solution, but the syntax is such that I’m not sure how I’d use it.</p>

**Comment by cschleiden** [comment]
<p>The basic idea is to set the keys as environment secrets and then always use the same name. Then you can use your <code>workflow_dispatch</code> input as the name of the environment, and the right set of secrets will be made available to the workflow:</p>
<pre><code class="lang-auto">  jobs:
    deploy:
      environment: ${{ github.event.inputs.env_to_promote_to }}
      steps:
        # This secret is named `AWS_ACCESS_KEY_ID` in _every_ environment
        - run: ./do-something ${{ secrets.AWS_ACCESS_KEY_ID }}</code></pre>

**Comment by mellis481** [comment]
<p><a class="mention" href="/cschleiden">@cschleiden</a>  When you say “environment”, what are you referring to? Is there a concept of environments in Github or the secrets module (which I’m unfamiliar with) where you can use the same secret ID for different values?</p>

**Comment by cschleiden** [comment, marked as answer]
<p>Yes, please see: <a href="https://docs.github.com/en/actions/reference/environments" rel="noopener nofollow ugc">Environments - GitHub Docs</a></p>
<p>And specifically this section: <a href="https://docs.github.com/en/actions/reference/environments#environment-secrets" rel="noopener nofollow ugc">Environments - Secrets - GitHub Docs</a></p>

**Comment by mellis481** [comment]
<p><a class="mention" href="/cschleiden">@cschleiden</a> This appears to be exactly what I need.  That said, my repos are private and my organization has GitHub Enterprise.  I’m seeing some statements in the page you linked that I’m interpreting as conflicting and I’m unclear if, given my situation, environment secrets are an option.</p>
<p>In the purple box, it reads “Environments, environment protection rules, and environment secrets are available in public repositories for all products and in private repositories for GitHub Enterprise.”  But then in the <em>About environments</em> section, it says “Environment protection rules and environment secrets are only available on public repositories.”</p>
<p>Are you able to say if environment secrets work for private repos with GitHub Enterprise?</p>

**Comment by cschleiden** [comment]
<p><a class="mention" href="/mellis481">@mellis481</a> That appears to be a bug in our documentation :grimacing: We’ll get that fixed.</p>
<p>On an enterprise plan environments (and all associated features) are also available for private repositories.</p>

**Comment by mellis481** [comment]
<p><a class="mention" href="/cschleiden">@cschleiden</a> I’m just now realizing that it appears the environment secrets are at the <em>repository</em>-level not <em>organization</em>-wide.  Is that correct?  If so, then it would appear this solution is not going to be satisfactory for me because the environment-based secrets I need go across numerous repositories in my organization.</p>

**Comment by cschleiden** [comment]
<p>Yes, environment are repository scoped.</p>

**Comment by mellis481** [comment]
<p><a class="mention" href="/cschleiden">@cschleiden</a>  are organization-level secrets on any roadmap?</p>

**Comment by dradetsky** [comment]
<p><a class="mention" href="/cschleiden">@cschleiden</a> I too am interested in org-wide per-env secrets.</p>

**Comment by tpak** [comment]
<p>@ <a href="https://github.community/u/cschleiden">cschleiden</a> Is there any plan to bring this to Pro and not force small companies to upgrade to Enterprise? We’re boot-strapped and cash-poor. This seems like one that every single organization needs, not just enterprises and open source projects.</p>
